### ISSUE TITLE
CA2012 fires for ValueTasks that are explicitly discarded

### ISSUE THREAD

**Issue by alexrp** [dotnet/roslyn-analyzers#3617]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

CA2012

#### Repro steps

```csharp
System.Threading.Channels.Channel<int> chan = ...;
_ = chan.Writer.WriteAsync(42);
```

#### Expected behavior

No diagnostic since it's an explicit discard, not just a plain local assignment.

#### Actual behavior

`CA2012: ValueTask instances returned from method calls should be directly awaited, returned, or passed as an argument to another method call. Other usage, such as storing an instance into a local or a field, is likely an indication of a bug, as ValueTask instances must only ever be consumed once.`

**Comment by mavasani** [comment]
This looks to be a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/3573.
Can you please try with https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/3.3.0-beta1.20258.1?

**Comment by alexrp** [comment]
I hadn't noticed that issue, but it's not quite a duplicate anyway. That issue was just about the diagnostic message being inaccurate (and was resolved by adjusting it for this case). What I'm arguing here is that emitting the diagnostic at all makes no sense for this case; if I'm discarding the `ValueTask` explicitly (`_ = FooAsync()`) then I'm probably doing so for a good reason, e.g.:

* I know that the specific implementation of `FooAsync` (which is `virtual`) that's in use will always complete synchronously, and so `await`ing it is just a waste of cycles. (This is the case in the repro.)
* `FooAsync` may or may not complete asynchronously, but I don't actually care because I just want to call it for its side effects in a fire-and-forget fashion.

**Comment by mavasani** [comment]
Tagging @stephentoub

**Comment by stephentoub** [comment]
> emitting the diagnostic at all makes no sense for this case

It does, though.  Consider your second bullet point:

> FooAsync may or may not complete asynchronously, but I don't actually care because I just want to call it for its side effects in a fire-and-forget fashion.

And let's take Channel.WriteAsync as an example.  That implementation uses a single pooled object that it can reuse as the backing store for the ValueTask over and over and over... but only when the previous operation is done with it.  And the only way it can no that the previous operation is done with it is when GetResult on the awaiter is called to consume the ValueTask.  If you discard it as in your example, you will no longer use the pooled object, and all subsequent asynchronously completing operations will allocate a new instance, for the lifetime of that channel.

In general, the semantics of ValueTask are that you need to consume it in one way, shape, or form.  If you really don't need it, you can call AsTask() or Preserve(), and then if desired discard the result of those.

I'm worried that explicitly allowing discards here will actually push developers in very much the wrong direction.

**Comment by alexrp** [comment]
> That implementation uses a single pooled object that it can reuse as the backing store for the ValueTask over and over and over...

That might be the case for some implementations, but AFAICT not for this one: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Threading.Channels/src/System/Threading/Channels/SingleConsumerUnboundedChannel.cs#L348 (To be clear, that's the one I'm using in the situation that motivated me to open this issue.)

**Comment by stephentoub** [comment]
Why not use TryWrite in that case?

**Comment by alexrp** [comment]
It's been a while since I actually wrote that code but IIRC my reasoning at the time was that `WriteAsync` more clearly communicated the intent of the code; `TryWrite` suggests some sort of possibility of failure that isn't actually there.

**Comment by stephentoub** [comment]
And WriteAsync suggests you need to wait for the result but you're not :smile:

My point is, you're writing code here assuming you don't need to consume the ValueTask, but the ValueTask contract says you do.  If the argument is that a particular implementation based on looking at its source code demonstrates you don't actually have to in that particular case, then you can similarly look at the source code for TryWrite and see that it will only return true in this case.  And of course in both situations, the implementation could change in the future such that, as long as it continues to adhere to the rules around ValueTask, the assumptions about whether the operation completed synchronously or not or whether it used a pooled object or not might change.  Case in point, `async ValueTask` methods currently don't pool, but if someone using .NET 5 sets the relevant flag (https://devblogs.microsoft.com/dotnet/async-valuetask-pooling-in-net-5/), then they will be pooled, and any code making similar discard assumptions about the result will be wrong.

We could of course suppress the rule for the discard case.  But the current implementation goes out of its way to warn for discards:
https://github.com/dotnet/roslyn-analyzers/blob/2a091354a54e59fe8a1677ecdb09f51364e7da97/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/UseValueTasksCorrectly.cs#L166-L171
because it's considered to be misuse and in violation of the ValueTask contract.  I'd like more evidence that the rule is actually a problem before giving up on that.

As it stands, you have a bunch of options, all of which would stay in conformance:
- Use `TryWrite(item);` instead of `_ = WriteAsync(item);`
- Use `WriteAsync(item).Preserve();` instead of `_ WriteAsync(item);`
- Use `_ = WriteAsync(item).Preserve();` instead of `_ = WriteAsync(item);`
- Use `_ = WriteAsync(item).AsTask();` instead of `_ = WriteAsync(item);`

**Comment by alexrp** [comment]
Okay, I think I'm on the same page as you are now.

I was aware of `ValueTask` restrictions such as the fact that they can only be `await`ed once, that `GetAwaiter` can't be called before it's completed, and so on. But the notion that you're *expected* not to discard them (and that .NET 5 will make use of this) is totally new to me. I think that's why we came at this issue from completely different angles.

It seems like it would be worthwhile to update the documentation for `ValueTask` to make this expectation known, and probably also make the (somewhat terse) documentation for `Preserve` a bit more detailed as it's not immediately obvious what it exists for.

Anyhow, I agree that the current diagnostic is fine given this expectation.

**Comment by stephentoub** [comment]
Thanks, @alexrp. Yeah, we should definitely beef up the docs here.

**Comment by moaazassali** [comment]
I have the same issue, but I am not sure if what has been mentioned here applies to my code. I made a TCP server to process concurrent connections, so I can't `await` my HandleClient() method because I have to keep listening to incoming connections.

```
public async Task Start()
{
    TcpListener.Start();
    Console.WriteLine($"Server started. Listening on {ServerIP}:{ServerPort}");

    while (true)
    {
        try
        {
            TcpClient client = await TcpListener.AcceptTcpClientAsync();
            _ = HandleClient(client);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error accepting client: {ex.Message}");
        }
    }
}

private async ValueTask HandleClient(TcpClient client) 
{
    while (true) 
    {
         ...
         int bytes = await stream.ReadAsync(buffer, cts.Token);
         ...
         await stream.WriteAsync(response);
         ...
    }
}
```
What are my options here? `Preserve()` seems pointless because if it uses `Task` under the hood I might as well just make my method return `Task` instead of `ValueTask`.
